File: TarsEngine.FSharp.Core/Metascript/Services/MetascriptExecutor.fs
namespace TarsEngine.FSharp.Core.Metascript.Services

open System
open System.IO
open System.Text.Json
open System.Threading.Tasks
open Microsoft.Extensions.Logging
open TarsEngine.FSharp.Core.Metascript
open TarsEngine.FSharp.Core.Metascript.Services

/// <summary>
/// Implementation of the IMetascriptExecutor interface.
/// </summary>
type MetascriptExecutor(logger: ILogger<MetascriptExecutor>, metascriptService: IMetascriptService) =
    /// <summary>
    /// Executes a metascript asynchronously.
    /// </summary>
    /// <param name="metascriptPath">The path to the metascript file.</param>
    /// <param name="parameters">Optional parameters to pass to the metascript.</param>
    /// <returns>The result of the metascript execution.</returns>
    member _.ExecuteMetascriptAsync(metascriptPath: string, ?parameters: obj) =
        Task.Run(fun () ->
            try
                logger.LogInformation("Executing metascript: {MetascriptPath}", metascriptPath)

                // Check if the file exists
                if not (File.Exists(metascriptPath)) then
                    logger.LogError("Metascript file not found: {MetascriptPath}", metascriptPath)
                    {
                        Success = false
                        Output = None
                        ErrorMessage = Some $"Metascript file not found: {metascriptPath}"
                    }
                else
                    // Read the metascript content
                    let metascriptContent = File.ReadAllText(metascriptPath)

                    // REAL METASCRIPT EXECUTION - NO SIMULATION
                    logger.LogInformation("Executing metascript file with REAL F# execution: {MetascriptPath}", metascriptPath)

                    // Use the MetascriptService for real execution
                    let metascriptService = MetascriptService(logger :?> ILogger<MetascriptService>)
                    let executionResult = (metascriptService.ExecuteMetascriptAsync(metascriptContent)).Result

                    // Return real execution result
                    {
                        Success = true
                        Output = Some (executionResult.ToString())
                        ErrorMessage = None
                    }
            with
            | ex ->
                logger.LogError(ex, "Error executing metascript: {MetascriptPath}", metascriptPath)
                {
                    Success = false
                    Output = None
                    ErrorMessage = Some ex.Message
                }
        )
    
    interface IMetascriptExecutor with
        member this.ExecuteMetascriptAsync(metascriptPath, ?parameters) =
            this.ExecuteMetascriptAsync(metascriptPath, ?parameters = parameters)

File: archive/v1/parked_csharp/Legacy_CSharp_Projects/TarsEngine/Metascripts/MetascriptExecutor.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using TarsEngine.Services.Interfaces;

namespace TarsEngine.Metascripts
{
    /// <summary>
    /// Implementation of the <see cref="IMetascriptExecutor"/> interface.
    /// </summary>
    public class MetascriptExecutor : IMetascriptExecutor
    {
        private readonly ILogger<MetascriptExecutor> _logger;
        private readonly IMetascriptService _metascriptService;

        /// <summary>
        /// Initializes a new instance of the <see cref="MetascriptExecutor"/> class.
        /// </summary>
        /// <param name="logger">The logger.</param>
        /// <param name="metascriptService">The metascript service.</param>
        public MetascriptExecutor(
            ILogger<MetascriptExecutor> logger,
            IMetascriptService metascriptService)
        {
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
            _metascriptService = metascriptService ?? throw new ArgumentNullException(nameof(metascriptService));
        }

        /// <inheritdoc/>
        public async Task<MetascriptExecutionResult> ExecuteMetascriptAsync(string metascriptPath, object? parameters = null)
        {
            try
            {
                _logger.LogInformation("Executing metascript: {MetascriptPath}", metascriptPath);

                // Check if the file exists
                if (!File.Exists(metascriptPath))
                {
                    _logger.LogError("Metascript file not found: {MetascriptPath}", metascriptPath);
                    return new MetascriptExecutionResult
                    {
                        Success = false,
                        ErrorMessage = $"Metascript file not found: {metascriptPath}"
                    };
                }

                // Read the metascript content
                var metascriptContent = await File.ReadAllTextAsync(metascriptPath);

                // Convert parameters to JSON if provided
                string parametersJson = string.Empty;
                if (parameters != null)
                {
                    parametersJson = JsonSerializer.Serialize(parameters);
                }

                // Execute the metascript
                var result = await _metascriptService.ExecuteMetascriptAsync(metascriptContent);

                // Convert the result to a MetascriptExecutionResult
                return new MetascriptExecutionResult
                {
                    Success = true,
                    Output = result?.ToString()
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error executing metascript: {MetascriptPath}", metascriptPath);
                return new MetascriptExecutionResult
                {
                    Success = false,
                    ErrorMessage = ex.Message
                };
            }
        }
    }
}
